File: Shadow samurei/rust/kern/src/bewertungen.rs
//! Bewertungen einzelner Antworten in `data/bewertungen.json`.
//!
//! Eine Bewertung ist `1` (hilfreich), `-1` (nicht hilfreich) oder `0`
//! (ohne Bewertung). Alte Dateien (`data/feedback.json`) und alte
//! englische Feldnamen werden weiterhin gelesen.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::path::{Path, PathBuf};

use crate::zeit;

/// Längste gespeicherte Text­länge für Frage und Antwort.
const HOECHSTLAENGE_TEXT: usize = 1000;

/// Gibt den Anzeigetext einer Bewertung zurück.
pub fn bewertungstext(bewertung: i32) -> &'static str {
    match bewertung {
        1 => "Hilfreich",
        -1 => "Nicht hilfreich",
        _ => "Ohne Bewertung",
    }
}

/// Kürzt einen Text auf die Höchstlänge.
fn kuerze(text: &str) -> String {
    if text.chars().count() <= HOECHSTLAENGE_TEXT {
        text.to_string()
    } else {
        text.chars().take(HOECHSTLAENGE_TEXT).collect()
    }
}

/// Eine einzelne Bewertung.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BewertungsEintrag {
    /// Zeitstempel im ISO-Format.
    pub zeitstempel: String,
    /// Modell, das die Antwort erzeugt hat.
    pub modell: String,
    /// Gestellte Frage.
    pub frage: String,
    /// Gegebene Antwort.
    pub antwort: String,
    /// Bewertung: 1, -1 oder 0.
    pub bewertung: i32,
    /// Freie Markierungen.
    pub markierungen: Vec<String>,
}

impl Default for BewertungsEintrag {
    fn default() -> Self {
        Self {
            zeitstempel: zeit::jetzt_iso(),
            modell: "unbekannt".into(),
            frage: String::new(),
            antwort: String::new(),
            bewertung: 0,
            markierungen: Vec::new(),
        }
    }
}

impl BewertungsEintrag {
    /// Erzeugt einen Eintrag; die Texte werden gekürzt.
    pub fn neu(modell: &str, frage: &str, antwort: &str, bewertung: i32) -> Self {
        Self {
            zeitstempel: zeit::jetzt_iso(),
            modell: if modell.is_empty() {
                "unbekannt".to_string()
            } else {
                modell.to_string()
            },
            frage: kuerze(frage),
            antwort: kuerze(antwort),
            bewertung: bewertung.clamp(-1, 1),
            markierungen: Vec::new(),
        }
    }

    /// Liest einen Eintrag aus – auch alten – JSON-Daten.
    pub fn aus_wert(daten: &Value) -> Self {
        let text = |deutsch: &str, englisch: &str, ersatz: &str| -> String {
            match daten.get(deutsch).or_else(|| daten.get(englisch)) {
                Some(Value::String(wert)) if !wert.is_empty() => wert.clone(),
                Some(Value::Null) | None => ersatz.to_string(),
                Some(anderer) => anderer.to_string(),
            }
        };
        let bewertung = match daten.get("bewertung").or_else(|| daten.get("rating")) {
            Some(Value::Number(zahl)) => zahl.as_f64().unwrap_or(0.0) as i32,
            Some(Value::String(wert)) => wert.trim().parse::<i32>().unwrap_or(0),
            Some(Value::Bool(true)) => 1,
            _ => 0,
        };
        let markierungen = match daten.get("markierungen").or_else(|| daten.get("tags")) {
            Some(Value::Array(liste)) => liste
                .iter()
                .map(|wert| match wert {
                    Value::String(text) => text.clone(),
                    anderer => anderer.to_string(),
                })
                .collect(),
            Some(Value::String(einzel)) if !einzel.is_empty() => vec![einzel.clone()],
            _ => Vec::new(),
        };
        let zeitstempel = {
            let gelesen = text("zeitstempel", "ts", "");
            if gelesen.is_empty() {
                zeit::jetzt_iso()
            } else {
                gelesen
            }
        };
        Self {
            zeitstempel,
            modell: text("modell", "model", "unbekannt"),
            frage: kuerze(&text("frage", "prompt", "")),
            antwort: kuerze(&text("antwort", "response", "")),
            bewertung: bewertung.clamp(-1, 1),
            markierungen,
        }
    }

    /// Gibt den Eintrag als JSON-Wert zurück.
    pub fn als_wert(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }

    /// Gibt den Anzeigetext der Bewertung zurück.
    pub fn text(&self) -> &'static str {
        bewertungstext(self.bewertung)
    }

    /// Gibt eine kurze, lesbare Zeitangabe zurück.
    pub fn kurzzeit(&self) -> String {
        zeit::kurzzeit(&self.zeitstempel)
    }
}

/// Kennzahlen über alle Bewertungen.
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
pub struct BewertungsZusammenfassung {
    /// Anzahl aller Bewertungen.
    pub gesamt: usize,
    /// Anzahl der hilfreichen Bewertungen.
    pub positiv: usize,
    /// Anzahl der nicht hilfreichen Bewertungen.
    pub negativ: usize,
    /// Anzahl der Einträge ohne Bewertung.
    pub neutral: usize,
    /// Anteil hilfreicher Bewertungen zwischen 0 und 1.
    pub anteil: f64,
}

/// Speichert und wertet Bewertungen aus.
#[derive(Debug, Clone)]
pub struct BewertungsSpeicher {
    pfad: PathBuf,
    alter_pfad: Option<PathBuf>,
}

impl BewertungsSpeicher {
    /// Öffnet den Speicher unter dem angegebenen Pfad.
    pub fn neu(pfad: &Path) -> Self {
        Self {
            pfad: pfad.to_path_buf(),
            alter_pfad: pfad.parent().map(|ordner| ordner.join("feedback.json")),
        }
    }

    /// Öffnet den Speicher unter `data/bewertungen.json` des Projekts.
    pub fn standardpfad() -> Self {
        Self::neu(&crate::pfade::datendatei("bewertungen.json"))
    }

    /// Gibt den Pfad der Bewertungsdatei zurück.
    pub fn pfad(&self) -> &Path {
        &self.pfad
    }

    /// Gibt alle Bewertungen zurück.
    pub fn hole_alle(&self) -> Vec<BewertungsEintrag> {
        let mut quellen = vec![self.pfad.clone()];
        quellen.extend(self.alter_pfad.clone());
        for pfad in quellen {
            let Ok(inhalt) = std::fs::read_to_string(&pfad) else {
                continue;
            };
            return match serde_json::from_str::<Value>(&inhalt) {
                Ok(Value::Array(liste)) => liste
                    .iter()
                    .filter(|wert| wert.is_object())
                    .map(BewertungsEintrag::aus_wert)
                    .collect(),
                _ => Vec::new(),
            };
        }
        Vec::new()
    }

    /// Schreibt alle Bewertungen in die Datei.
    fn schreibe(&self, eintraege: &[BewertungsEintrag]) {
        let liste: Vec<Value> = eintraege.iter().map(BewertungsEintrag::als_wert).collect();
        if let Ok(text) = serde_json::to_string_pretty(&Value::Array(liste)) {
            let _ = crate::pfade::schreibe_atomar(&self.pfad, &text);
        }
    }

    /// Legt eine Bewertung an und speichert sie.
    pub fn fuege_hinzu(&self, eintrag: BewertungsEintrag) -> BewertungsEintrag {
        let mut eintraege = self.hole_alle();
        eintraege.push(eintrag.clone());
        self.schreibe(&eintraege);
        eintrag
    }

    /// Legt eine Bewertung aus JSON-Daten an.
    pub fn fuege_wert_hinzu(&self, rohdaten: &Value) -> BewertungsEintrag {
        self.fuege_hinzu(BewertungsEintrag::aus_wert(rohdaten))
    }

    /// Gibt die letzten `anzahl` Bewertungen zurück.
    pub fn hole_letzte(&self, anzahl: usize) -> Vec<BewertungsEintrag> {
        if anzahl == 0 {
            return Vec::new();
        }
        let alle = self.hole_alle();
        let beginn = alle.len().saturating_sub(anzahl);
        alle[beginn..].to_vec()
    }

    /// Berechnet die Kennzahlen über alle Bewertungen.
    pub fn zusammenfassung(&self) -> BewertungsZusammenfassung {
        let eintraege = self.hole_alle();
        let positiv = eintraege.iter().filter(|e| e.bewertung > 0).count();
        let negativ = eintraege.iter().filter(|e| e.bewertung < 0).count();
        let neutral = eintraege.len() - positiv - negativ;
        let bewertet = positiv + negativ;
        BewertungsZusammenfassung {
            gesamt: eintraege.len(),
            positiv,
            negativ,
            neutral,
            anteil: if bewertet == 0 {
                0.0
            } else {
                positiv as f64 / bewertet as f64
            },
        }
    }

    /// Löscht alle Bewertungen.
    pub fn leere(&self) {
        self.schreibe(&[]);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn speicher() -> (tempfile::TempDir, BewertungsSpeicher) {
        let ordner = tempfile::tempdir().expect("Temporärordner");
        let speicher = BewertungsSpeicher::neu(&ordner.path().join("data/bewertungen.json"));
        (ordner, speicher)
    }

    #[test]
    fn leerer_speicher_gibt_nullen() {
        let (_ordner, speicher) = speicher();
        assert!(speicher.hole_alle().is_empty());
        assert_eq!(
            speicher.zusammenfassung(),
            BewertungsZusammenfassung::default()
        );
    }

    #[test]
    fn bewertungen_werden_gezaehlt() {
        let (_ordner, speicher) = speicher();
        speicher.fuege_hinzu(BewertungsEintrag::neu("modell", "Frage", "Antwort", 1));
        speicher.fuege_hinzu(BewertungsEintrag::neu("modell", "Frage", "Antwort", 1));
        speicher.fuege_hinzu(BewertungsEintrag::neu("modell", "Frage", "Antwort", -1));
        speicher.fuege_hinzu(BewertungsEintrag::neu("modell", "Frage", "Antwort", 0));
        let kennzahlen = speicher.zusammenfassung();
        assert_eq!(kennzahlen.gesamt, 4);
        assert_eq!(kennzahlen.positiv, 2);
        assert_eq!(kennzahlen.negativ, 1);
        assert_eq!(kennzahlen.neutral, 1);
        assert!((kennzahlen.anteil - 2.0 / 3.0).abs() < 1e-9);
        assert_eq!(speicher.hole_letzte(2).len(), 2);
    }

    #[test]
    fn lange_texte_werden_gekuerzt() {
        let (_ordner, speicher) = speicher();
        let eintrag = speicher.fuege_hinzu(BewertungsEintrag::neu(
            "",
            &"ä".repeat(1500),
            &"b".repeat(1500),
            5,
        ));
        assert_eq!(eintrag.frage.chars().count(), 1000);
        assert_eq!(eintrag.antwort.chars().count(), 1000);
        assert_eq!(eintrag.bewertung, 1);
        assert_eq!(eintrag.modell, "unbekannt");
    }

    #[test]
    fn alte_feldnamen_werden_gelesen() {
        let (_ordner, speicher) = speicher();
        let eintrag = speicher.fuege_wert_hinzu(&json!({
            "ts": "2026-01-02T08:30:00", "model": "alt", "prompt": "Warum?",
            "response": "Deshalb.", "rating": "-1", "tags": "web"
        }));
        assert_eq!(eintrag.modell, "alt");
        assert_eq!(eintrag.frage, "Warum?");
        assert_eq!(eintrag.antwort, "Deshalb.");
        assert_eq!(eintrag.bewertung, -1);
        assert_eq!(eintrag.markierungen, vec!["web".to_string()]);
        assert_eq!(eintrag.text(), "Nicht hilfreich");
        assert_eq!(eintrag.kurzzeit(), "02.01. 08:30");
    }

    #[test]
    fn alte_datei_wird_gelesen() {
        let ordner = tempfile::tempdir().expect("Temporärordner");
        let datenordner = ordner.path().join("data");
        std::fs::create_dir_all(&datenordner).expect("Ordner");
        std::fs::write(
            datenordner.join("feedback.json"),
            "[{\"model\": \"alt\", \"rating\": 1}]",
        )
        .expect("schreiben");
        let speicher = BewertungsSpeicher::neu(&datenordner.join("bewertungen.json"));
        assert_eq!(speicher.hole_alle().len(), 1);
        assert_eq!(speicher.zusammenfassung().positiv, 1);
    }

    #[test]
    fn leeren_entfernt_alles() {
        let (_ordner, speicher) = speicher();
        speicher.fuege_hinzu(BewertungsEintrag::default());
        speicher.leere();
        assert!(speicher.hole_alle().is_empty());
    }
}

File: Shadow samurei py/kern/bewertungen.py
"""Bewertungen einzelner Antworten in data/bewertungen.json.

Eine Bewertung ist 1 (hilfreich), -1 (nicht hilfreich) oder 0
(ohne Bewertung). Alte Dateien (data/feedback.json) und alte
englische Feldnamen werden weiterhin gelesen.
"""

import json
import os

from .pfade import datendatei, schreibe_atomar
from .zeit import jetzt_iso, kurzzeit

#: Längste gespeicherte Textlänge für Frage und Antwort.
HOECHSTLAENGE_TEXT = 1000


def bewertungstext(bewertung):
    """Gibt den Anzeigetext einer Bewertung zurück."""
    if bewertung == 1:
        return "Hilfreich"
    elif bewertung == -1:
        return "Nicht hilfreich"
    else:
        return "Ohne Bewertung"


def _kuerze(text):
    """Kürzt einen Text auf die Höchstlänge."""
    if len(text) <= HOECHSTLAENGE_TEXT:
        return text
    return text[:HOECHSTLAENGE_TEXT]


class BewertungsEintrag:
    """Eine einzelne Bewertung."""

    def __init__(
        self,
        zeitstempel=None,
        modell="unbekannt",
        frage="",
        antwort="",
        bewertung=0,
        markierungen=None,
    ):
        self.zeitstempel = zeitstempel if zeitstempel else jetzt_iso()
        self.modell = modell
        self.frage = frage
        self.antwort = antwort
        self.bewertung = bewertung
        self.markierungen = markierungen if markierungen is not None else []

    @classmethod
    def neu(cls, modell, frage, antwort, bewertung):
        """Erzeugt einen Eintrag; die Texte werden gekürzt."""
        return cls(
            zeitstempel=jetzt_iso(),
            modell="unbekannt" if not modell else modell,
            frage=_kuerze(frage),
            antwort=_kuerze(antwort),
            bewertung=max(-1, min(1, bewertung)),
            markierungen=[],
        )

    @classmethod
    def aus_wert(cls, daten):
        """Liest einen Eintrag aus – auch alten – JSON-Daten."""
        def text(deutsch, englisch, ersatz=""):
            wert = daten.get(deutsch, daten.get(englisch)) if isinstance(daten, dict) else None
            if isinstance(wert, str) and wert:
                return wert
            if wert is None:
                return ersatz
            return str(wert)

        bewertung = 0
        bewertung_wert = daten.get("bewertung", daten.get("rating")) if isinstance(daten, dict) else None
        if isinstance(bewertung_wert, bool) and bewertung_wert:
            bewertung = 1
        elif isinstance(bewertung_wert, (int, float)) and not isinstance(bewertung_wert, bool):
            bewertung = int(bewertung_wert)
        elif isinstance(bewertung_wert, str):
            try:
                bewertung = int(bewertung_wert.strip())
            except ValueError:
                pass

        markierungen = []
        mark_wert = daten.get("markierungen", daten.get("tags")) if isinstance(daten, dict) else None
        if isinstance(mark_wert, list):
            markierungen = [
                m if isinstance(m, str) else str(m) for m in mark_wert
            ]
        elif isinstance(mark_wert, str) and mark_wert:
            markierungen = [mark_wert]

        zeitstempel = text("zeitstempel", "ts", "")
        if not zeitstempel:
            zeitstempel = jetzt_iso()

        return cls(
            zeitstempel=zeitstempel,
            modell=text("modell", "model", "unbekannt"),
            frage=_kuerze(text("frage", "prompt", "")),
            antwort=_kuerze(text("antwort", "response", "")),
            bewertung=max(-1, min(1, bewertung)),
            markierungen=markierungen,
        )

    def als_wert(self):
        """Gibt den Eintrag als JSON-Wert zurück."""
        return {
            "zeitstempel": self.zeitstempel,
            "modell": self.modell,
            "frage": self.frage,
            "antwort": self.antwort,
            "bewertung": self.bewertung,
            "markierungen": self.markierungen,
        }

    def text(self):
        """Gibt den Anzeigetext der Bewertung zurück."""
        return bewertungstext(self.bewertung)

    def kurzzeit(self):
        """Gibt eine kurze, lesbare Zeitangabe zurück."""
        return kurzzeit(self.zeitstempel)


class BewertungsZusammenfassung:
    """Kennzahlen über alle Bewertungen."""

    def __init__(self, gesamt=0, positiv=0, negativ=0, neutral=0, anteil=0.0):
        self.gesamt = gesamt
        self.positiv = positiv
        self.negativ = negativ
        self.neutral = neutral
        self.anteil = anteil

    def als_wert(self):
        return {
            "gesamt": self.gesamt,
            "positiv": self.positiv,
            "negativ": self.negativ,
            "neutral": self.neutral,
            "anteil": self.anteil,
        }


class BewertungsSpeicher:
    """Speichert und wertet Bewertungen aus."""

    def __init__(self, pfad=None):
        if pfad is None:
            pfad = datendatei("bewertungen.json")
        self.pfad = pfad
        ordner = os.path.dirname(pfad)
        self._alter_pfad = (
            os.path.join(ordner, "feedback.json") if ordner else "feedback.json"
        )

    @classmethod
    def standardpfad(cls):
        """Öffnet den Speicher unter data/bewertungen.json des Projekts."""
        return cls(datendatei("bewertungen.json"))

    def hole_alle(self):
        """Gibt alle Bewertungen zurück."""
        quellen = [self.pfad, self._alter_pfad]
        for pfad in quellen:
            try:
                with open(pfad, "r", encoding="utf-8") as f:
                    inhalt = f.read()
            except FileNotFoundError:
                continue
            try:
                liste = json.loads(inhalt)
            except json.JSONDecodeError:
                return []
            if isinstance(liste, list):
                return [
                    BewertungsEintrag.aus_wert(w).als_wert()
                    for w in liste
                    if isinstance(w, dict)
                ]
            return []
        return []

    def _schreibe(self, eintraege):
        """Schreibt alle Bewertungen in die Datei."""
        liste = [
            e.als_wert() if isinstance(e, BewertungsEintrag) else e
            for e in eintraege
        ]
        text = json.dumps(liste, indent=2, ensure_ascii=False)
        schreibe_atomar(self.pfad, text)

    def fuege_hinzu(self, daten):
        """Legt eine Bewertung an und speichert sie."""
        if isinstance(daten, BewertungsEintrag):
            eintrag = daten
        else:
            eintrag = BewertungsEintrag.aus_wert(daten)
        eintraege = [BewertungsEintrag.aus_wert(e) for e in self.hole_alle()]
        eintraege.append(eintrag)
        self._schreibe(eintraege)
        return eintrag.als_wert()

    def fuege_wert_hinzu(self, rohdaten):
        """Legt eine Bewertung aus JSON-Daten an."""
        return self.fuege_hinzu(rohdaten)

    def hole_letzte(self, anzahl):
        """Gibt die letzten anzahl Bewertungen zurück."""
        if anzahl == 0:
            return []
        alle = self.hole_alle()
        return alle[-anzahl:]

    def zusammenfassung(self):
        """Berechnet die Kennzahlen über alle Bewertungen."""
        eintraege = self.hole_alle()
        positiv = sum(1 for e in eintraege if e.get("bewertung", 0) > 0)
        negativ = sum(1 for e in eintraege if e.get("bewertung", 0) < 0)
        neutral = len(eintraege) - positiv - negativ
        bewertet = positiv + negativ
        anteil = positiv / bewertet if bewertet > 0 else 0.0
        return BewertungsZusammenfassung(
            gesamt=len(eintraege),
            positiv=positiv,
            negativ=negativ,
            neutral=neutral,
            anteil=anteil,
        ).als_wert()

    def leere(self):
        """Löscht alle Bewertungen."""
        self._schreibe([])

    @staticmethod
    def bewertungstext(bewertung):
        """Gibt den Anzeigetext einer Bewertung zurück."""
        return bewertungstext(bewertung)
